src/vec_poly.h

```diff
@@ -0,0 +1,234 @@
+/* -*- mode: c++ -*- */
+/*
+ * Copyright 2017-2018 Scality
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef __NTTEC_VEC_POLY_H__
+#define __NTTEC_VEC_POLY_H__
+
+#include "gf_base.h"
+#include "vec_vector.h"
+
+namespace nttec {
+namespace vec {
+
+/** A vector of size \f$n\f$ represents a polynomial \f$P(X)\f$ of degree
+ * \f$(n-1)\f$
+ *
+ * Vector: \f$v[0], v[1], ..., v[n-1]\f$ represents the polynomial:
+ *  \f$P(X):= v[0] + v[1] X + ... + v[n-1] X^n\f$
```

Comment:
Wasn't that already the case for `nttec::Polynomial`? Is this class superseding `nttec::Polynomial` (lot of code seems duplicated otherwise, like `eval` is exactly the same code in both classes)?

I think I don't get the difference between `nettc::vec::Poly` and `nttec::Polynomial`: what's the difference? When do use one over the other?

Reply:
`nttec::Polynomial` is defined over std::map
`nttec::vec::Poly` is extended on vec::Vector, i.e. over an array
The two classes have same APIs that are used for decoding.

Reply:
As seen with you, `nttec::Vec::Poly` is not really a vec (so it doesn't belong to the namespace `vec`), but it is a polynomial that uses a vector internally (like `nttec::Polynomial` is not a map, but a polynomial using a `std::map`

A more logical representation would be:
- an abstract base class `Polynomial` that defines the interface of a polynomial (eval, derive, …)
- a child class SparsePolynomial (that would correspond to the current `nttec::Polynomial`)
- a child class FullPolynomial (that would be your new `nttec::vec::Poly`).

Also, inheritance represents an IS-A relations, and you can't really say that a polynomial IS-A vector (nor IS-A map) so it would make more sense to use composition here (having a member Vec in your class).

As there require heavy change to your PR, let's forget it for now (but I'll create an issue to rework that).

Reply:
I forgot that FFT will perform on `vec::Poly`. While input and output of FFT are vec::Vector (and vec::Buffer). Hence we should take into account this constraint.